File: backtracking/79_word_search.cc
#include <iostream>
#include <string>
#include <vector>

const std::vector<std::vector<int>> directions = {
    {1, 0}, {-1, 0}, {0, 1}, {0, -1}};

bool Dfs(std::vector<std::vector<char>>& board,
         std::vector<std::vector<bool>>& is_visited, const std::string& word,
         int i, int j, int k) {
  if (i < 0 || i >= board.size() || j < 0 || j >= board[0].size() ||
      k >= word.size()) {
    return false;
  }
  if (is_visited[i][j] || board[i][j] != word[k]) {
    return false;
  }
  if (board[i][j] == word[k] && k == word.size() - 1) {
    return true;
  }
  is_visited[i][j] = true;
  for (const auto& direction : directions) {
    if (Dfs(board, is_visited, word, i + direction[0], j + direction[1],
            k + 1)) {
      return true;
    }
  }
  is_visited[i][j] = false;
  return false;
}

bool Exist(std::vector<std::vector<char>>& board, const std::string& word) {
  int m = board.size();
  int n = board[0].size();
  std::vector<std::vector<bool>> is_visited(m, std::vector<bool>(n, false));
  for (int i = 0; i < m; ++i) {
    for (int j = 0; j < n; ++j) {
      if (Dfs(board, is_visited, word, i, j, 0)) {
        return true;
      }
    }
  }
  return false;
}

int main(int argc, char* argv[]) {
  std::vector<std::vector<char>> board = {
      {'A', 'B', 'C', 'E'}, {'S', 'F', 'C', 'S'}, {'A', 'D', 'E', 'E'}};
  std::string word = "ABCCED";
  std::cout << Exist(board, word) << std::endl;
  return 0;
}

File: backtracking/79_word_search.py
from typing import List


def exist(board: List[List[str]], word: str) -> bool:
    is_exist = False
    if not board or not board[0]:
        return is_exist

    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    m, n = len(board), len(board[0])

    result = []
    is_visited = [[0 for i in range(n)] for i in range(m)]
    word_index = 0

    def backtracking(row: int, col: int) -> None:
        nonlocal is_exist, word_index, is_visited
        if row < 0 or row >= m or col < 0 or col >= n or word_index >= len(word):
            return

        if is_visited[row][col]:
            return

        if board[row][col] != word[word_index]:
            return

        if word_index == len(word) - 1:
            is_exist = True
            return

        is_visited[row][col] = 1
        result.append(board[row][col])
        for direction in directions:
            row = row + direction[0]
            col = col + direction[1]
            word_index += 1
            backtracking(row, col)
            word_index -= 1
            row = row - direction[0]
            col = col - direction[1]
        is_visited[row][col] = 0
        result.pop()

    for i in range(m):
        for j in range(n):
            backtracking(i, j)

    return is_exist


board = [["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]]
word = "ABCCED"
print(exist(board, word))
